### ISSUE TITLE
Too many heap sections capturing check(8, "DiffAlg") on armhf

### ISSUE THREAD

**Issue by d-torrance** [Macaulay2/M2#1728]
From PPA builds on Ubuntu [18.04](https://launchpadlibrarian.net/512788986/buildlog_ubuntu-bionic-armhf.macaulay2_1.16.99+git202012211524-0ppa202012211755~ubuntu18.04.1_BUILDING.txt.gz) and [20.04](https://launchpadlibrarian.net/512800594/buildlog_ubuntu-focal-armhf.macaulay2_1.16.99+git202012211524-0ppa202012211755~ubuntu20.04.1_BUILDING.txt.gz) armf last night:

```m2
 -- capturing check(8, "DiffAlg") from source:
   DiffAlg.m2:1324:1:                                                       Too many heap sections: Increase MAXHINCR or MAX_HEAP_SECTS
Aborted (core dumped)
make[4]: *** [Makefile:39: check-DiffAlg] Error 134
make[4]: Leaving directory '/<<BUILDDIR>>/macaulay2-1.16.99+git202012211524/M2/Macaulay2/packages'
make[3]: *** [Makefile:14: check-in-packages] Error 2
make[3]: Leaving directory '/<<BUILDDIR>>/macaulay2-1.16.99+git202012211524/M2/Macaulay2'
make[2]: *** [GNUmakefile:229: check-in-Macaulay2] Error 2
make[2]: Leaving directory '/<<BUILDDIR>>/macaulay2-1.16.99+git202012211524/M2'
```

The PPA builds for Ubuntu 20.10 and 21.04 armhf finished without problems.

It's interesting that we didn't get to the `test #8 of package DiffAlg failed` message.  I suppose this is because of the switch to `capture`, and the parent `M2` aborted, not some child one that it called just for this test?

**Comment by DanGrayson** [comment]
Yes, if the program crashes, then it doesn't get to capture the next test.

Another thing about capturing is that all the garbage memory from the previous tests is still there.

**Comment by mahrud** [comment]
Interesting. Let me think about this a bit. If it's unavoidable, we can disable that for the build.

> Another thing about capturing is that all the garbage memory from the previous tests is still there.

I added garbage collection after capturing in #1716

**Comment by mahrud** [comment]
@d-torrance does this happen on top of #1716 as well?

**Comment by d-torrance** [comment]
> @d-torrance does this happen on top of #1716 as well?

I'll test it out and see.

**Comment by mahrud** [comment]
Also, if it is just a few tests that break things, we can use `-* no-capture-flag *-` to disable capturing for those tests only.

**Comment by d-torrance** [comment]
> > @d-torrance does this happen on top of #1716 as well?
> 
> I'll test it out and see.

Yes, it does.  On a Raspberry Pi on top of #1716:

```m2
i1 : scan(20, i -> (print i; check_8 "DiffAlg"))
0
 -- capturing check(8, "DiffAlg")                                            -- -1.0382 seconds elapsed
1
 -- capturing check(8, "DiffAlg")                                            -- 1.25852 seconds elapsed
2
 -- capturing check(8, "DiffAlg")                                           Too many heap sections: Increase MAXHINCR or MAX_HEAP_SECTS
Aborted
```
This was no difference running from `development` -- it crashed on the 3rd attempt there,t oo.

**Comment by d-torrance** [comment]
> Also, if it is just a few tests or examples that crash, we can use `-* no-capture-flag *-` to disable capturing for those tests or examples only.

Good to know -- thanks!

**Comment by mahrud** [comment]
This is probably a memory leak, right?

**Comment by d-torrance** [comment]
> This is probably a memory leak, right?

I would assume so.

**Comment by mahrud** [comment]
This is actually cool, because it means we might be able to detect memory leaks using capture!

**Comment by d-torrance** [comment]
`-* no-capture-flag *-` works like a charm!  I'll open a PR adding that to this test.

**Comment by mahrud** [comment]
Feel free to push to my ongoing capture PR instead.

**Comment by d-torrance** [comment]
> Feel free to push to my ongoing capture PR instead.

Oops -- I didn't see this until after I'd opened a new PR.  I'll just leave the new one open.

**Comment by d-torrance** [comment]
There was a similar issue two tests later on a [powerpc build of the Debian 1.17.1 package](https://buildd.debian.org/status/fetch.php?pkg=macaulay2&arch=powerpc&ver=1.17.1%2Bds-1&stamp=1610933689&raw=0):

```m2
 -- capturing check(10, "DiffAlg")                                          Too many heap sections: Increase MAXHINCR or MAX_HEAP_SECTS
Aborted
make[4]: *** [Makefile:39: check-DiffAlg] Error 134
```

**Comment by mahrud** [comment]
Hmm, okay, would you like to reopen this to track work on the memory leak? In the meantime, adding the flag should help.

**Comment by d-torrance** [comment]
> Hmm, okay, would you like to reopen this to track work on the memory leak?

Probably not a bad idea.  I don't have the ability to re-open it, though.

> In the meantime, adding the flag should help.

I'll submit a PR in a moment.

**Comment by mahrud** [comment]
@d-torrance do you happen know what are the engine level functions that DiffAlg uses?

**Comment by d-torrance** [comment]
> @d-torrance do you happen know what are the engine level functions that DiffAlg uses?

It definitely uses `rawMatrixDiff`. Not sure if that's causing the leak.

**Comment by mahrud** [comment]
@mikestillman could that function have a memory leak?

**Comment by d-torrance** [comment]
Even after switching away from capture, we still ran out of memory during `check(7, "Diffalg")` in a [powerpc Debian build](https://buildd.debian.org/status/fetch.php?pkg=macaulay2&arch=powerpc&ver=1.17.1%2Bds-2&stamp=1611609924&raw=0):

```m2
-- running   check(7, "DiffAlg")                                           
 ulimit -c unlimited; ulimit -t 700; ulimit -m 850000; ulimit -s 8192; ulimit -n 512;  cd /tmp/M2-1648137-0/15-rundir/; GC_MAXIMUM_HEAP_SIZE=400M "/<<PKGBUILDDIR>>/M2/usr-dist/powerpc-Linux-Debian-unstable/bin/M2-binary" -q --int --no-randomize --no-readline --silent --stop --print-width 129 -e 'needsPackage("DiffAlg",Reload=>true,FileName=>"/<<PKGBUILDDIR>>/M2/Macaulay2/packages/DiffAlg.m2")' <"/tmp/M2-1648137-0/14.m2" >>"/tmp/M2-1648137-0/14.tmp" 2>&1
/tmp/M2-1648137-0/14.tmp:0:1: (output file) error: Macaulay2 exited with status code 1
/tmp/M2-1648137-0/14.m2:0:1: (input file)
M2: *** Error 1
 -- 103.417 seconds elapsed
 -- running   check(8, "DiffAlg")                                            -- 105.877 seconds elapsed
 -- running   check(9, "DiffAlg")                                            -- 3.90946 seconds elapsed
 -- running   check(10, "DiffAlg")                                           -- 6.43395 seconds elapsed
 -- running   check(11, "DiffAlg")                                           -- 1.2534 seconds elapsed
 -- running   check(12, "DiffAlg")                                           -- 1.36369 seconds elapsed
 -- running   check(13, "DiffAlg")                                           -- 3.04357 seconds elapsed
 -- running   check(14, "DiffAlg")                                           -- 14.9448 seconds elapsed
DiffAlg.m2:1316:1: error:
 --      ------x  - }
 --      926327 0  
 --      -----------
 -- 
 -- o3 : List
 -- 
 -- i4 : assert( (w^(diff (L_0)) + (L_0) ^ (diff w))#"f" == 0)
 -- 
 -- 
 --  *** out of memory trying to allocate 131108 bytes, exiting ***
 -- 
../m2/debugging.m2:20:6:(1):[9]: error: test(s) #7 of package DiffAlg failed.
```

**Comment by DanGrayson** [comment]
Let's just comment out the test, and inform the authors by email, so they may propose an alternative test if coverage is reduced.

**Comment by d-torrance** [comment]
This seems to have been fixed by #2344.  The `DiffAlg` tests are passing on armhf builds in both Ubuntu [18.04](https://launchpadlibrarian.net/573297102/buildlog_ubuntu-bionic-armhf.macaulay2_1.19.1.1+git202112072115-0ppa202112091114~ubuntu18.04.1_BUILDING.txt.gz) and [20.04](https://launchpadlibrarian.net/573323666/buildlog_ubuntu-focal-armhf.macaulay2_1.19.1.1+git202112072115-0ppa202112091114~ubuntu20.04.1_BUILDING.txt.gz):

```m2
/<<BUILDDIR>>/macaulay2-1.19.1.1+git202112072115/M2/usr-dist/arm-Linux-Ubuntu-20.04/bin/M2 -q --no-preload --stop --silent -e "needsPackage(\"DiffAlg\",LoadDocumentation=>true,DebuggingMode=>true); check(DiffAlg,UserMode=>false,Verbose=>true); exit 0" 
 -- capturing check(0, "DiffAlg")                                            -- 0.82931 seconds elapsed
 -- capturing check(1, "DiffAlg")                                            -- 7.32423 seconds elapsed
 -- capturing check(2, "DiffAlg")                                            -- 2.96873 seconds elapsed
 -- capturing check(3, "DiffAlg")                                            -- 1.5157 seconds elapsed
 -- capturing check(4, "DiffAlg")                                            -- 22.8743 seconds elapsed
 -- capturing check(5, "DiffAlg")                                            -- 1.10419 seconds elapsed
 -- capturing check(6, "DiffAlg")                                            -- 21.9638 seconds elapsed
 -- capturing check(7, "DiffAlg")                                            -- 213.443 seconds elapsed
 -- capturing check(8, "DiffAlg")                                            -- 223.525 seconds elapsed
 -- capturing check(9, "DiffAlg")                                            -- 8.37546 seconds elapsed
 -- capturing check(10, "DiffAlg")                                           -- 18.7507 seconds elapsed
 -- capturing check(11, "DiffAlg")                                           -- 2.46165 seconds elapsed
 -- capturing check(12, "DiffAlg")                                           -- 2.54908 seconds elapsed
 -- capturing check(13, "DiffAlg")                                           -- 7.44313 seconds elapsed
 -- capturing check(14, "DiffAlg")                                           -- 41.6737 seconds elapsed
```

Closing.
